service/src/main/java/io/camunda/service/UserTaskServices.java

```diff
@@ -40,11 +58,176 @@ public UserTaskServices withAuthentication(final Authentication authentication)
+            ? allProcessInstanceIds
+            : originalFilter.processInstanceKeys(),
+        originalFilter.processDefinitionKeys(),
+        allScopeKeys,
+        originalFilter.candidateUsers(),
+        originalFilter.candidateGroups(),
+        originalFilter.created(),
+        originalFilter.completed(),
+        originalFilter.canceled(),
+        originalFilter.creationDateFilter(),
+        originalFilter.completionDateFilter(),
+        originalFilter.dueDateFilter(),
+        originalFilter.followUpDateFilter(),
+        originalFilter.variableFilters(),
+        originalFilter.tenantIds());
+  }
+
   public SearchQueryResult<UserTaskEntity> search(
       final Function<Builder, ObjectBuilder<UserTaskQuery>> fn) {
     return search(SearchQueryBuilders.userTaskSearchQuery(fn));
   }
+
+  private void processVariableFilter(
+      final VariableValueFilter variableValueFilter,
+      final VariableServices variableServices,
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey,
+      final boolean firstIteration) {
+    final VariableFilter.Builder variableFilterBuilder = new VariableFilter.Builder();
+    variableFilterBuilder.variable(variableValueFilter);
+
+    final SearchQueryPage searchQueryPage =
+        new SearchQueryPage.Builder().size(MAX_ELASTICSEARCH_PAGE_SIZE).build();
+
+    final VariableQuery variableQueryBuilder;
+    if (firstIteration) {
+      variableQueryBuilder =
+          VariableQuery.of(b -> b.filter(variableFilterBuilder.build()).page(searchQueryPage));
+    } else {
+      variableFilterBuilder.processInstanceKeys(
+          new ArrayList<>(scopeKeyMapByProcessInstanceKey.keySet()));
+      variableFilterBuilder.variable(variableValueFilter);
+      variableQueryBuilder =
+          VariableQuery.of(b -> b.filter(variableFilterBuilder.build()).page(searchQueryPage));
+    }
+
+    final var searchResults =
+        variableServices.withAuthentication(authentication).search(variableQueryBuilder);
+
+    final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey =
+        searchResults.items().stream()
+            .collect(
+                Collectors.groupingBy(
+                    VariableEntity::processInstanceKey,
+                    Collectors.mapping(VariableEntity::scopeKey, Collectors.toSet())));
+
+    if (firstIteration) {
+      initializeScopeKeyMap(
+          scopeKeyMapByProcessInstanceKey, currentScopeKeyMapByProcessInstanceKey);
+    } else {
+      updateScopeKeyMap(scopeKeyMapByProcessInstanceKey, currentScopeKeyMapByProcessInstanceKey);
+    }
+  }
+
+  private void initializeScopeKeyMap(
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey,
+      final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey) {
+    for (final Map.Entry<Long, Set<Long>> entry :
+        currentScopeKeyMapByProcessInstanceKey.entrySet()) {
+      final Long processInstanceId = entry.getKey();
```

Comment:
:x: Please change to `processInstanceKey`, we should use the same terms for the same things all over the place to avoid confusions.